README.md

# Vec Filter

Vec Filter is a Rust library to filter a vector of structs based on a query string. It allows you to specify filter conditions on struct fields with a simple query syntax. The library provides a custom derive macro, `Filterable`, to make your structs filterable with ease.

## Table of Contents

-   [Installation](#installation)
-   [Usage](#usage)
    -   [Deriving Filterable](#deriving-filterable)
    -   [Query Syntax](#query-syntax)
    -   [Example](#example)
-   [License](#license)

## Installation

Add the following to your `Cargo.toml`:

```toml
[dependencies]
vec_filter = "0.1.0"
```

## Usage

### Deriving Filterable

To make your struct filterable, derive the `Filterable` trait:

```rust
use vec_filter::Filterable;

#[derive(Debug, Clone, Filterable)]
struct Person {
    name: String,
    age: u32,
}
```

### Query Syntax

This section describes the string query format that can be parsed by the library. The string queries are composed of a series of expressions that represent operations, values, and logical connectors. Expressions can be combined using parentheses to create more complex queries.

#### Operations

*   `==`: Equals
*   `!=`: Not equals
*   `>`: Greater than
*   `>=`: Greater than or equal to
*   `<`: Less than
*   `<=`: Less than or equal to
*   `contains`: Contains substring
*   `startswith`: Starts with substring
*   `endswith`: Ends with substring
*   `regexmatch`: Matches regex pattern
*   `in`: Checks if a value is in a list of values

#### Values

Values can be of the following types:

*   String: Enclosed in double quotes, e.g. `"hello"`
*   Integer: A sequence of digits, e.g. `42`
*   List of strings: Enclosed in square brackets, separated by commas, e.g. `["apple", "banana", "cherry"]`
*   List of integers: Enclosed in square brackets, separated by commas, e.g. `[1, 2, 3]`

#### Logical Connectors

*   `&&`: Logical AND
*   `||`: Logical OR
*   `!`: Logical NOT (used with parentheses)

#### Examples

*   `field1 == "value1"`: field1 equals "value1"
*   `(field1 != "value1") && (field2 > 42)`: field1 is not equal to "value1" and field2 is greater than 42
*   `(field1 contains "substr") || (field2 < 10)`: field1 contains the substring "substr" or field2 is less than 10
*   `!(field1 >= 100)`: field1 is not greater than or equal to 100
*   `field1 in ["apple", "banana"]`: field1 is either "apple" or "banana"
*   `((field1 == "value1") || (field1 == "value2")) && (field2 <= 20)`: field1 is either "value1" or "value2", and field2 is less than or equal to 20

#### Parentheses
To set the precedence of the operations, use parentheses. Expressions enclosed in parentheses will be evaluated first. If you want to group conditions, you can use parentheses to create more complex queries.

For example:

* `((age >= 25) && (interests in ["hiking"])) || (name == "Alice")`: This query will match items where the age is greater than or equal to 25 and the interests contain "hiking", or the name is equal to "Alice".
* `(name == "Alice") || ((age < 30) && (interests in ["cooking"]))`: This query will match items where the name is equal to "Alice", or the age is less than 30 and the interests contain "cooking".

By using parentheses, you can build complex queries that combine multiple conditions with different levels of precedence to achieve precise filtering.

#### Notes

*   Spaces are allowed between elements but are not required.
*   The field names must be valid Rust identifiers.
*   The parser is case-sensitive; field names and string values must match the case exactly.

### Example

Let's see how to filter a vector of `Person` structs using Vec Filter.

```rust
use vec_filter::{Filterable, AST};
use vec_filter::Value;

#[derive(Debug, Clone, PartialEq, Filterable)]
struct Person {
    name: String,
    age: u32,
}

fn main() {
    let people = vec![
        Person { name: "Alice".to_string(), age: 30 },
        Person { name: "Bob".to_string(), age: 40 },
        Person { name: "Charlie".to_string(), age: 50 },
    ];

    let ast = AST::And(
        Box::new(AST::Equals {
            field: "name",
            value: Value::String("Alice".to_string()),
        }),
        Box::new(AST::NotEquals {
            field: "age",
            value: Value::Int(30),
        }),
    );

    let filtered_people = ast.apply(&people);
    println!("{:?}", filtered_people); // []
}
```

In this example, we want to find all people with the name "Alice" and age not equal to 30. The query is represented using an `AST` value. The `apply` method is then called with the `people` vector, and the filtered result is printed.

## License

This project is licensed under the Apache 2.0 License.


vec_filter/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.2.2 (2023-03-27)

Add new AST options

### New Features

 - <csr-id-2b24b0212229813378056bd5c77abc86ff026d07/> Add multiple input strings for parse_query benchmark
 - <csr-id-5eb205ddc9b9191c0baf07ef98f3858947a03364/> Add microbenchmarks for parse_query and apply

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 3 commits contributed to the release over the course of 2 calendar days.
 - 2 days passed between releases.
 - 2 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - WIP ([`beba1be`](https://github.com/andymac4182/vec_filter/commit/beba1be2d75a184657abaf5d52110d1680e02000))
    - Add multiple input strings for parse_query benchmark ([`2b24b02`](https://github.com/andymac4182/vec_filter/commit/2b24b0212229813378056bd5c77abc86ff026d07))
    - Add microbenchmarks for parse_query and apply ([`5eb205d`](https://github.com/andymac4182/vec_filter/commit/5eb205ddc9b9191c0baf07ef98f3858947a03364))
</details>

## 0.2.1 (2023-03-22)

Add doc strings for AST, Move internal methods to extension trait

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release vec_filter v0.2.1 ([`fabe7c8`](https://github.com/andymac4182/vec_filter/commit/fabe7c823ab4aa45421089c8f6774edf9c6301a8))
    - Add doc strings for AST, Move internal methods to extension trait ([`4111150`](https://github.com/andymac4182/vec_filter/commit/4111150285b0cb7657a2129fcc5427d01dcfa0aa))
</details>

## 0.2.0 (2023-03-22)

Test

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 5 commits contributed to the release.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release vec_filter_derive v0.1.1, vec_filter v0.2.0 ([`d58f77d`](https://github.com/andymac4182/vec_filter/commit/d58f77d1da5e94f1b113f053ab98a52ae35906ce))
    - Adjusting changelogs prior to release of vec_filter_derive v0.1.1, vec_filter v0.2.0 ([`92e6ae6`](https://github.com/andymac4182/vec_filter/commit/92e6ae6a3fa68cf02697b8776106944f3a856a05))
    - Update changelog ([`aed0969`](https://github.com/andymac4182/vec_filter/commit/aed0969236aa011b155bd796fa7d4ba6746e2249))
    - Move tests ([`6cc93ef`](https://github.com/andymac4182/vec_filter/commit/6cc93ef95897d0ce1cdcf994d2b29082ff025d9b))
    - Adjusting changelogs prior to release of vec_filter v0.2.0 ([`2b6deef`](https://github.com/andymac4182/vec_filter/commit/2b6deef2ea9635ca5e1b3e1806a67d030faf2be8))
</details>

## 0.1.0 (2023-03-22)

Initial Release

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 12 commits contributed to the release.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release vec_filter v0.1.0 ([`8a3b195`](https://github.com/andymac4182/vec_filter/commit/8a3b1950a561ddb2b04b19402fcafac664c50d6b))
    - Release vec_filter_derive v0.1.0, vec_filter v0.1.0 ([`142c71b`](https://github.com/andymac4182/vec_filter/commit/142c71bda5082da35ec0ad6bf0bbee3221195b3b))
    - Release vec_filter_derive v0.1.0, vec_filter v0.1.0 ([`cf0d389`](https://github.com/andymac4182/vec_filter/commit/cf0d3891a0e2c84e1eaac708b40118b79b576c79))
    - Release vec_filter_derive v0.1.0, vec_filter v0.1.0 ([`900d870`](https://github.com/andymac4182/vec_filter/commit/900d870e0acebcbda6727adf412da5e84f7f5ea4))
    - Update changelog ([`1aef4ec`](https://github.com/andymac4182/vec_filter/commit/1aef4eca2c61ddba0f2a30ccb70025b6ccfff01e))
    - Adjusting changelogs prior to release of vec_filter_derive v0.1.0, vec_filter v0.1.0 ([`f2e589a`](https://github.com/andymac4182/vec_filter/commit/f2e589a6087d7a1c12ca40a6106c029a1ab417eb))
    - Adjusting changelogs prior to release of vec_filter_derive v0.1.0, vec_filter v0.1.0 ([`554e202`](https://github.com/andymac4182/vec_filter/commit/554e2029bc0de103f32d7bbe44f670c838f37290))
    - Adjusting changelogs prior to release of vec_filter_derive v0.1.0, vec_filter v0.1.0 ([`a79bc03`](https://github.com/andymac4182/vec_filter/commit/a79bc03c25dfaf1a19d5c3aae4a5eac0013ab4b6))
    - Add CHANGELOG ([`07262ae`](https://github.com/andymac4182/vec_filter/commit/07262ae1008fe4614a8283220c639cc655c5248b))
    - Fix up cargo.toml ([`01a867e`](https://github.com/andymac4182/vec_filter/commit/01a867eb166309e370450546d680d23dd224e3f4))
    - Add GT, LT, GTE, LTE ([`d2dadc6`](https://github.com/andymac4182/vec_filter/commit/d2dadc6ee37e53e22af5244ca75474668bc250ec))
    - Initial Commit ([`9a93f13`](https://github.com/andymac4182/vec_filter/commit/9a93f13386c34f2b80e8b237a81233a49383ffd7))
</details>



vec_filter/tests/strings.rs
use vec_filter::{parse_query, Filterable, StructProperties, Value, AST};

use std::fmt;
use std::str::FromStr;

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[derive(Debug, Clone, PartialEq, Filterable)]
    pub struct Person {
        pub name: String,
        pub age: u32,
        pub interests: Vec<String>,
    }

    #[rstest]
    #[case::name_equals_alice("name == \"Alice\"", vec![0])]
    #[case::name_not_equals_alice("name != \"Alice\"", vec![1, 2])]
    #[case::age_equals_30("age == 30", vec![0])]
    #[case::age_not_equals_30("age != 30", vec![1, 2])]
    #[case::interests_contains_reading("interests in [\"reading\"]", vec![0])]
    #[case::interests_contains_cooking("interests in [\"cooking\"]", vec![1])]
    #[case::interests_contains_hiking("interests in [\"hiking\"]", vec![0, 2])]
    #[case::name_equals_alice_and_age_equals_30("(name == \"Alice\") && (age == 30)", vec![0])]
    #[case::name_equals_alice_or_name_equals_bob("(name == \"Alice\") || (name == \"Bob\")", vec![0, 1])]
    #[case::name_in_alice_or_bob("name in [\"Alice\",\"Bob\"]", vec![0, 1])]
    #[case::interests_in_hiking("interests in \"hiking\"", vec![0, 2])]
    #[case::name_alice_bob_eve("(name == \"Alice\") || (name == \"Bob\") || (name == \"Eve\")", vec![0, 1])]
    #[case::interests_hiking_age_25("(interests in [\"hiking\"]) && (age == 25)", vec![2])]
    #[case::name_equals_alice_interests_hiking_age_20("((name == \"Alice\") && (interests in [\"hiking\"])) || (age == 20)", vec![0, 1])]
    #[case::interests_hiking_age_20_or_25("(interests in [\"hiking\"]) && ((age == 20) || (age == 25))", vec![2])]
    #[case::age_greater_than_25("age > 25", vec![0])]
    #[case::age_less_than_25("age < 25", vec![1])]
    #[case::age_greater_than_equal_to_25("age >= 25", vec![0, 2])]
    #[case::age_less_than_equal_to_25("age <= 25", vec![1, 2])]
    #[case::name_startswith_al("name startswith \"Al\"", vec![0])]
    #[case::name_endswith_ce("name endswith \"ce\"", vec![0])]
    #[case::name_regexmatch_alice_bob("name regexmatch \"^(Alice|Bob)$\"", vec![0, 1])]
    #[case::not_operation("!(age == 30)", vec![1, 2])]
    fn test_filtering(#[case] input: &str, #[case] expected_indices: Vec<usize>) {
        let alice = Person {
            name: "Alice".to_string(),
            age: 30,
            interests: vec!["reading".to_string(), "hiking".to_string()],
        };
        let bob = Person {
            name: "Bob".to_string(),
            age: 20,
            interests: vec!["swimming".to_string(), "cooking".to_string()],
        };
        let carol = Person {
            name: "Carol".to_string(),
            age: 25,
            interests: vec!["hiking".to_string(), "painting".to_string()],
        };

        let people = vec![alice, bob, carol];

        let ast = parse_query(input).unwrap_or_else(|err| {
            panic!("Failed to parse input '{}': {:?}", input, err);
        });

        let filtered_people: Vec<Person> = ast.apply(&people);
        let expected_people: Vec<Person> = expected_indices
            .iter()
            .map(|index| people[*index].clone())
            .collect();

        assert_eq!(filtered_people, expected_people);
    }

    #[rstest]
    #[case::equals_operation("age == 30", AST::Equals { field: PersonProperties::age, value: Value::Int(30) })]
    #[case::not_equals_operation("age != 30", AST::NotEquals { field: PersonProperties::age, value: Value::Int(30) })]
    #[case::contains_operation("interests in [\"hiking\"]", AST::In { field: PersonProperties::interests, value: Value::VecString(vec!["hiking".to_string()]) })]
    #[case::and_operation("(age == 30) && (interests in [\"hiking\"])", AST::And(
        Box::new(AST::Equals { field: PersonProperties::age, value: Value::Int(30) }),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::VecString(vec!["hiking".to_string()]) })
    ))]
    #[case::or_operation("(age == 30) || (interests in [\"hiking\"])", AST::Or(
        Box::new(AST::Equals { field: PersonProperties::age, value: Value::Int(30) }),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::VecString(vec!["hiking".to_string()]) })
    ))]
    #[case::greater_than_operation("age > 25", AST::GreaterThan { field: PersonProperties::age, value: Value::Int(25) })]
    #[case::greater_than_or_equal_operation("age >= 25", AST::GreaterThanOrEqual { field: PersonProperties::age, value: Value::Int(25) })]
    #[case::less_than_operation("age < 25", AST::LessThan { field: PersonProperties::age, value: Value::Int(25) })]
    #[case::less_than_or_equal_operation("age <= 25", AST::LessThanOrEqual { field: PersonProperties::age, value: Value::Int(25) })]
    #[case::interests_contains_operation("interests in \"hiking\"", AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })]
    #[case::not_operation("!(age == 30)", AST::Not(Box::new(AST::Equals { field: PersonProperties::age, value: Value::Int(30) })))]
    #[case::complex_and_operation("(age > 25) && (interests in \"hiking\")", AST::And(
        Box::new(AST::GreaterThan { field: PersonProperties::age, value: Value::Int(25) }),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })
    ))]
    #[case::complex_or_operation("(age < 25) || (interests in \"hiking\")", AST::Or(
        Box::new(AST::LessThan { field: PersonProperties::age, value: Value::Int(25) }),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })
    ))]
    #[case::complex_mixed_operations("((age > 20) && (age < 30)) || (interests in \"hiking\")", AST::Or(
        Box::new(AST::And(
            Box::new(AST::GreaterThan { field: PersonProperties::age, value: Value::Int(20) }),
            Box::new(AST::LessThan { field: PersonProperties::age, value: Value::Int(30) })
        )),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })
    ))]
    #[case::name_equals_operation("name == \"Alice\"", AST::Equals { field: PersonProperties::name, value: Value::String("Alice".to_string()) })]
    #[case::name_not_equals_operation("name != \"Alice\"", AST::NotEquals { field: PersonProperties::name, value: Value::String("Alice".to_string()) })]
    #[case::name_contains_operation("name contains \"A\"", AST::Contains { field: PersonProperties::name, value: Value::String("A".to_string()) })]
    #[case::and_and_operation("((age > 20) && (age < 30)) && (interests in \"hiking\")", AST::And(
        Box::new(AST::And(
            Box::new(AST::GreaterThan { field: PersonProperties::age, value: Value::Int(20) }),
            Box::new(AST::LessThan { field: PersonProperties::age, value: Value::Int(30) })
        )),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })
    ))]
    #[case::or_or_operation("((age > 20) && (age < 30)) || ((name in \"A\") || (interests in \"hiking\"))", AST::Or(
        Box::new(AST::And(
            Box::new(AST::GreaterThan { field: PersonProperties::age, value: Value::Int(20) }),
            Box::new(AST::LessThan { field: PersonProperties::age, value: Value::Int(30) })
        )),
        Box::new(AST::Or(
            Box::new(AST::In { field: PersonProperties::name, value: Value::String("A".to_string()) }),
            Box::new(AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })
        ))
    ))]
    #[case::complex_nested_operations("(!(age > 20) && (name contains \"A\")) || (interests in \"hiking\")", AST::Or(
        Box::new(AST::And(
            Box::new(AST::Not(Box::new(AST::GreaterThan { field: PersonProperties::age, value: Value::Int(20) }))),
            Box::new(AST::Contains { field: PersonProperties::name, value: Value::String("A".to_string()) })
        )),
        Box::new(AST::In { field: PersonProperties::interests, value: Value::String("hiking".to_string()) })
    ))]
    fn test_parse_query_ast_output(#[case] input: &str, #[case] expected_ast: AST<PersonProperties>) {
        let ast = parse_query::<PersonProperties>(input).unwrap();
        assert_eq!(ast, expected_ast);
    }

    #[derive(Clone, Debug)]
    struct DummyProperties;

    impl StructProperties for DummyProperties {
        fn valid_fields() -> Vec<&'static str> {
            vec![]
        }

        fn get_value_type(&self) -> Value {
            Value::String("".to_string())
        }
    }

    impl FromStr for DummyProperties {
        type Err = ();

        fn from_str(_s: &str) -> Result<Self, Self::Err> {
            Ok(DummyProperties)
        }
    }

    impl fmt::Display for DummyProperties {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(f, "[]")
        }
    }
    /*
    #[rstest::rstest]
    #[case("name == 123", "Expected a string for comparison with '=='")]
    #[case("age == \"Alice\"", "Expected a number for comparison with '=='")]
    #[case("interests in 123", "Expected a string or array for the 'in' operator")]
    #[case("name startswith 123", "Expected a string for the 'startswith' operator")]
    #[case("name endswith [\"Alice\"]", "Expected a string for the 'endswith' operator")]
    #[case("name regexmatch 123", "Expected a string for the 'regexmatch' operator")]
    fn test_error_handling(#[case] input: &str, #[case] expected_error: &str) {
            let result = parse_query(input);

            match result {
                Ok(_) => panic!("Expected an error, but got Ok(_)"),
                Err(nom::Err::Error(e)) => {
                    assert_eq!(e.input, expected_error);
                    // You can also assert the error kind if necessary.
                }
                Err(nom::Err::Failure(e)) => panic!("Expected an error, but got Failure: {:?}", e),
                Err(nom::Err::Incomplete(_)) => panic!("Expected an error, but got Incomplete"),
            }
        }
        */
}


vec_filter/tests/parse_errors.rs
use vec_filter::{parse_query, ASTValidationError, CompatibilityError, Filterable, Value, AST};

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[derive(Debug, Clone, PartialEq, Filterable)]
    pub struct Person {
        pub name: String,
        pub age: u32,
        pub interests: Vec<String>,
    }

    #[rstest]
    #[rstest]
    #[case::invalid_syntax("field1 == 'value'", ASTValidationError::InvalidSyntax)]
    #[case::invalid_field("field1 == \"value\"", ASTValidationError::InvalidField { field_name: "field1".to_string() })]
    #[case::invalid_comparison("age == \"Alice\"", ASTValidationError::CompatibilityError(CompatibilityError {
        ast: AST::Equals {
            field: PersonProperties::age,
            value: Value::String("Alice".to_string()),
        },
        field: PersonProperties::age,
        provided_value: Value::String("Alice".to_string()),
        valid_values: vec![Value::Int(i32::default())],
    }))]
    #[case::invalid_and_operator(
        "age > 25 && (name == \"Alice\"",
        ASTValidationError::InvalidSyntax
    )]
    #[case::invalid_or_operator(
        "age > 25 || (name == \"Alice\")",
        ASTValidationError::InvalidSyntax
    )]
    #[case::invalid_parentheses(
        "((age > 25) && (name == \"Alice\"",
        ASTValidationError::InvalidSyntax
    )]
    #[case::invalid_value_type("interests == \"reading\"", ASTValidationError::CompatibilityError(CompatibilityError {
        ast: AST::Equals {
            field: PersonProperties::interests,
            value: Value::String("reading".to_string()),
        },
        field: PersonProperties::interests,
        provided_value: Value::String("reading".to_string()),
        valid_values: vec![Value::VecString(vec![])],
    }))]
    #[case::nested_and_or(
        "age > 25 || (name == \"Alice\" && (interests in [\"reading\"]))",
        ASTValidationError::InvalidSyntax
    )]
    #[case::missing_operator(
        "(name == \"Alice\") (name == \"Bob\")",
        ASTValidationError::InvalidSyntax
    )]
    #[case::incorrect_and("age > 25 & (name == \"Alice\")", ASTValidationError::InvalidSyntax)]
    #[case::incorrect_or("age > 25 | (name == \"Alice\")", ASTValidationError::InvalidSyntax)]
    #[case::unmatched_parentheses(
        "((age > 25) && (name == \"Alice\")",
        ASTValidationError::InvalidSyntax
    )]
    #[case::missing_parentheses("age > 25 && name == \"Alice\"", ASTValidationError::InvalidSyntax)]
    //    #[case::extra_parentheses("((age > 25) && ((name == \"Alice\")))", ASTValidationError::InvalidSyntax)]
    #[case::mix_or_and_no_brackets(
        "(age > 25) && (name == \"Alice\") || age == 3",
        ASTValidationError::InvalidSyntax
    )]
    fn parse_query_error_handling(
        #[case] input: &str,
        #[case] expected_error: ASTValidationError<PersonProperties>,
    ) {
        let result = parse_query::<PersonProperties>(input);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err()[0], expected_error);
    }
}


example/src/main.rs
use vec_filter::{parse_query, Filterable};

#[derive(Debug, Clone, PartialEq, Filterable)]
pub struct Person {
    pub name: String,
    pub age: u32,
    pub interests: Vec<String>,
}

fn main() {
    let input = "(age > 25) && (name == \"Alice\") || age == 3";
    let result = parse_query(input);
    println!("AST: {:?}", result.as_ref().unwrap());

    let people = vec![
        Person {
            name: "Alice".to_string(),
            age: 30,
            interests: vec!["reading".to_string(), "hiking".to_string()],
        },
        Person {
            name: "Bob".to_string(),
            age: 25,
            interests: vec!["swimming".to_string(), "cooking".to_string()],
        },
        Person {
            name: "Charlie".to_string(),
            age: 35,
            interests: vec!["hiking".to_string(), "painting".to_string()],
        },
    ];

    match result {
        Ok(ast) => {
            let filtered_people: Vec<Person> = ast.apply(&people);
            println!("Filtered people: {:?}", filtered_people);
        }
        Err(e) => println!("Error: {:?}", e),
    }
}


vec_filter/src/ast.rs
use crate::{StructMatcher, StructProperties};
use std::convert::TryInto;
use std::fmt::Debug;
use std::str::FromStr;

/// An enumeration of Abstract Syntax Tree (AST) nodes representing various query operations.
/// P: The generic type parameter representing the field type in the AST.
#[derive(Debug, PartialEq, Clone)]
pub enum AST<P> {
    /// Represents an equality operation: field == value.
    Equals {
        field: P,
        value: Value,
    },

    /// Represents an inequality operation: field != value.
    NotEquals {
        field: P,
        value: Value,
    },

    /// Represents a containment operation: field in value.
    In {
        field: P,
        value: Value,
    },

    /// Represents a containment operation: field contains value.
    Contains {
        field: P,
        value: Value,
    },

    /// Represents a greater-than operation: field > value.
    GreaterThan {
        field: P,
        value: Value,
    },

    /// Represents a less-than operation: field < value.
    LessThan {
        field: P,
        value: Value,
    },

    /// Represents a greater-than-or-equal-to operation: field >= value.
    GreaterThanOrEqual {
        field: P,
        value: Value,
    },

    /// Represents a less-than-or-equal-to operation: field <= value.
    LessThanOrEqual {
        field: P,
        value: Value,
    },

    /// Represents a starts-with operation: field starts with value.
    StartsWith {
        field: P,
        value: Value,
    },

    /// Represents an ends-with operation: field ends with value.
    EndsWith {
        field: P,
        value: Value,
    },

    /// Represents a regex-match operation: field matches regex pattern.
    RegexMatch {
        field: P,
        value: Value,
    },

    InvalidField {
        field_name: String,
    },

    /// Represents a logical AND operation between two AST nodes.
    And(Box<AST<P>>, Box<AST<P>>),

    /// Represents a logical OR operation between two AST nodes.
    Or(Box<AST<P>>, Box<AST<P>>),

    Not(Box<AST<P>>),
}

impl<P> AST<P> {
    pub fn apply<F: StructMatcher<P> + Clone>(&self, items: &[F]) -> Vec<F> {
        items
            .iter()
            .filter(|item| item.matches_ast(self))
            .cloned()
            .collect()
    }
}

#[derive(Debug, PartialEq, PartialOrd, Clone)]
pub enum Value {
    String(String),
    Int(i32),
    VecString(Vec<String>),
    VecInt(Vec<i32>),
}

impl Value {
    pub fn wrap<T>(t: T) -> Self
    where
        T: Into<Value>,
    {
        t.into()
    }
}

impl AsRef<Value> for Value {
    fn as_ref(&self) -> &Value {
        self
    }
}

impl From<String> for Value {
    fn from(val: String) -> Self {
        Value::String(val)
    }
}

impl From<u32> for Value {
    fn from(val: u32) -> Self {
        Value::Int(val.try_into().unwrap())
    }
}

impl From<i32> for Value {
    fn from(val: i32) -> Self {
        Value::Int(val)
    }
}

impl From<Vec<String>> for Value {
    fn from(val: Vec<String>) -> Self {
        Value::VecString(val)
    }
}

impl From<Vec<i32>> for Value {
    fn from(val: Vec<i32>) -> Self {
        Value::VecInt(val)
    }
}

fn valid_comparison_values<P>(ast: &AST<P>, value: &Value) -> Vec<Value> {
    match ast {
        AST::Equals { field: _, value: _ } => {
            vec![value.clone()]
        }
        AST::NotEquals { field: _, value: _ } => {
            vec![value.clone()]
        }
        AST::GreaterThan { field: _, value: _ } => {
            if let Value::Int(ref n) = value {
                vec![Value::Int(*n - 1)]
            } else {
                vec![]
            }
        }
        AST::LessThan { field: _, value: _ } => {
            if let Value::Int(ref n) = value {
                vec![Value::Int(*n + 1)]
            } else {
                vec![]
            }
        }
        AST::GreaterThanOrEqual { field: _, value: _ } => {
            if let Value::Int(ref n) = value {
                vec![value.clone(), Value::Int(*n - 1)]
            } else {
                vec![]
            }
        }
        AST::LessThanOrEqual { field: _, value: _ } => {
            if let Value::Int(ref n) = value {
                vec![value.clone(), Value::Int(*n + 1)]
            } else {
                vec![]
            }
        }
        AST::Contains { field: _, value: _ } => {
            if let Value::String(ref s) = value {
                vec![Value::String(s.clone())]
            } else {
                vec![]
            }
        }
        AST::StartsWith { field: _, value: _ } => {
            if let Value::String(ref s) = value {
                vec![Value::String(s.clone())]
            } else {
                vec![]
            }
        }
        AST::EndsWith { field: _, value: _ } => {
            if let Value::String(ref s) = value {
                vec![Value::String(s.clone())]
            } else {
                vec![]
            }
        }
        AST::RegexMatch { field: _, value: _ } => {
            if let Value::String(ref s) = value {
                vec![Value::String(s.clone())]
            } else {
                vec![]
            }
        }
        AST::In { field: _, value: _ } => match value {
            Value::String(_) | Value::VecString(_) => {
                vec![Value::String(String::default()), Value::VecString(vec![])]
            }
            Value::Int(_) | Value::VecInt(_) => {
                vec![Value::Int(i32::default()), Value::VecInt(vec![])]
            }
        },
        AST::And(left, right) => {
            let mut values = vec![];
            values.append(&mut valid_comparison_values(left, value));
            values.append(&mut valid_comparison_values(right, value));
            values
        }
        AST::Or(left, right) => {
            let mut values = vec![];
            values.append(&mut valid_comparison_values(left, value));
            values.append(&mut valid_comparison_values(right, value));
            values
        }
        AST::InvalidField { field_name: _ } | AST::Not { .. } => unimplemented!("This should never be called"),
    }
}

#[derive(Debug, PartialEq)]
pub struct CompatibilityError<P: StructProperties> {
    pub ast: AST<P>,
    pub field: P,
    pub provided_value: Value,
    pub valid_values: Vec<Value>,
}

impl<P: StructProperties> std::fmt::Display for CompatibilityError<P> {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(
            f,
            "Incompatible value for field {:?}. Operation: {:?}, Provided value: {:?}, valid value options: {:?}",
            self.field, self.ast, self.provided_value, self.valid_values
        )
    }
}

impl<P: StructProperties> std::error::Error for CompatibilityError<P> {}

pub fn is_compatible<P: StructProperties>(
    ast: &AST<P>,
    parsed_value: &Value,
) -> Result<(), CompatibilityError<P>>
where
    <P as FromStr>::Err: Debug,
{
    fn variants_match(a: &Value, b: &Value) -> bool {
        std::mem::discriminant(a) == std::mem::discriminant(b)
    }

    let (field, field_value_type) = match ast {
        AST::Equals { field, .. }
        | AST::NotEquals { field, .. }
        | AST::In { field, .. }
        | AST::GreaterThan { field, .. }
        | AST::LessThan { field, .. }
        | AST::GreaterThanOrEqual { field, .. }
        | AST::LessThanOrEqual { field, .. }
        | AST::StartsWith { field, .. }
        | AST::EndsWith { field, .. }
        | AST::RegexMatch { field, .. }
        | AST::Contains { field, .. } 
            => (field, field.get_value_type()),
        AST::And { .. }
        | AST::Or { .. }
        | AST::InvalidField { .. } 
        | AST::Not { .. }
            => unreachable!("This variant should not be handled"),
    };

    let valid_values = valid_comparison_values(ast, &field_value_type);

    if valid_values
        .iter()
        .any(|value| variants_match(value, parsed_value))
    {
        Ok(())
    } else {
        Err(CompatibilityError {
            ast: ast.clone(),
            field: field.clone(),
            provided_value: parsed_value.clone(),
            valid_values,
        })
    }
}


vec_filter/benches/microbenchmarks.rs
use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion, Throughput};
use vec_filter::{parse_query, Filterable};

#[derive(Debug, Clone, PartialEq, Filterable)]
pub struct Person {
    pub name: String,
    pub age: u32,
    pub interests: Vec<String>,
}

fn bench_parse_query(c: &mut Criterion) {
    let input_strings = vec![
        "name == \"Alice\"",
        "name != \"Alice\"",
        "age == 30",
        "age != 30",
        "interests in [\"reading\"]",
        "interests in [\"cooking\"]",
        "interests in [\"hiking\"]",
        "(name == \"Alice\") && (age == 30)",
        "(name == \"Alice\") || (name == \"Bob\")",
        "name in [\"Alice\",\"Bob\"]",
        "interests in \"hiking\"",
        "(name == \"Alice\") || (name == \"Bob\") || (name == \"Eve\")",
        "(interests in [\"hiking\"]) && (age == 25)",
        "((name == \"Alice\") && (interests in [\"hiking\"])) || (age == 20)",
        "(interests in [\"hiking\"]) && ((age == 20) || (age == 25))",
        "age > 25",
        "age < 25",
        "age >= 25",
        "age <= 25",
    ];
    let mut group = c.benchmark_group("parse_query");
    for query in input_strings.iter() {
        group.throughput(Throughput::Elements(1));
        group.bench_with_input(BenchmarkId::from_parameter(query), query, |b, &query| {
            b.iter(|| parse_query::<PersonProperties>(black_box(query)))
        });
    }
    group.finish();
}

fn bench_apply(c: &mut Criterion) {
    let alice = Person {
        name: "Alice".to_string(),
        age: 30,
        interests: vec!["reading".to_string(), "hiking".to_string()],
    };
    let bob = Person {
        name: "Bob".to_string(),
        age: 20,
        interests: vec!["swimming".to_string(), "cooking".to_string()],
    };
    let carol = Person {
        name: "Carol".to_string(),
        age: 25,
        interests: vec!["hiking".to_string(), "painting".to_string()],
    };

    let people = vec![alice, bob, carol];
    let query = "((name == \"Alice\") && (interests in [\"hiking\"])) || (age == 20)";
    let ast = parse_query(query).unwrap();

    let mut group = c.benchmark_group("apply");
    group.throughput(Throughput::Elements(3));
    group.bench_function("ips", |b| b.iter(|| ast.apply(black_box(&people))));
    group.finish();
}

criterion_group!(benches, bench_parse_query, bench_apply);
criterion_main!(benches);
